api/types/access_list.go

```diff
@@ -0,0 +1,299 @@
+		}
+	}
+
+	if a.Spec.Audit.Frequency == 0 {
+		return trace.BadParameter("audit frequency must be greater than 0")
+	}
+
+	if len(a.Spec.Grants.Roles) == 0 && len(a.Spec.Grants.Traits) == 0 {
+		return trace.BadParameter("grants must specify at least one role or trait")
+	}
+
+	for _, member := range a.Spec.Members {
+		if member.Name == "" {
+			return trace.BadParameter("member name is missing")
+		}
+
+		if member.Joined.IsZero() {
+			return trace.BadParameter("member %s joined is missing", member.Name)
+		}
+
+		if member.Expires.IsZero() {
+			return trace.BadParameter("member %s expires is missing", member.Name)
+		}
+
+		if member.Reason == "" {
+			return trace.BadParameter("member %s reason is missing", member.Name)
+		}
+
+		if member.AddedBy == "" {
+			return trace.BadParameter("member %s added by is missing", member.Name)
+		}
+	}
+
+	return nil
+}
+
+// GetOwners returns the list of owners from the access list.
+func (a *AccessList) GetOwners() []AccessListOwner {
+	return a.Spec.Owners
+}
+
+// GetAuditFrequency returns the audit frequency from the access list.
+func (a *AccessList) GetAuditFrequency() time.Duration {
+	return a.Spec.Audit.Frequency
+}
+
+// GetMembershipRequires returns the membership requires configuration from the access list.
+func (a *AccessList) GetMembershipRequires() AccessListRequires {
+	return a.Spec.MembershipRequires
+}
+
+// GetOwnershipRequires returns the ownership requires configuration from the access list.
+func (a *AccessList) GetOwnershipRequires() AccessListRequires {
+	return a.Spec.OwnershipRequires
+}
+
+// GetGrants returns the grants from the access list.
+func (a *AccessList) GetGrants() AccessListGrants {
+	return a.Spec.Grants
+}
+
+// GetMembers returns the members from the access list.
+func (a *AccessList) GetMembers() []AccessListMember {
+	return a.Spec.Members
+}
+
+// GetMetadata returns metadata. This is specifically for conforming to the Resource interface,
+// and should be removed when possible.
+func (a *AccessList) GetMetadata() Metadata {
+	return FromHeaderMetadata(a.Metadata)
+}
+
+// MatchSearch goes through select field values of a resource
+// and tries to match against the list of search values.
+func (a *AccessList) MatchSearch(values []string) bool {
+	fieldVals := append(utils.MapToStrings(a.GetAllLabels()), a.GetName())
+	return MatchSearch(fieldVals, values, nil)
+}
+
+func (a *AccessListAudit) UnmarshalJSON(data []byte) error {
+	var audit map[string]interface{}
+	if err := json.Unmarshal(data, &audit); err != nil {
+		return trace.Wrap(err)
+	}
+
+	var err error
+	a.Frequency, err = time.ParseDuration(fmt.Sprintf("%v", audit["frequency"]))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func (a *AccessListAudit) MarshalJSON() ([]byte, error) {
+	audit := map[string]interface{}{}
+	audit["frequency"] = a.Frequency.String()
+	data, err := json.Marshal(audit)
+	return data, trace.Wrap(err)
+}
```

Comment:
nit: imo custom marshal/unmarshal should always have test coverage. too easy to accidentally break them without realizing I think.